sqlcapture/capture.go

```diff
@@ -628,6 +650,45 @@ func (c *Capture) emitMessage(out *boilerplate.PullOutput, msg interface{}) erro
 	}
 }
 
+func (c *Capture) acknowledgeWorker(ctx context.Context, stream *boilerplate.PullOutput, replStream ReplicationStream) error {
+	for {
+		var request, err = stream.Recv()
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			return err
+		} else if err = request.Validate_(); err != nil {
+			return fmt.Errorf("validating request: %w", err)
+		}
+
+		switch {
+		case request.Acknowledge != nil:
+			if err := c.handleAcknowledgement(ctx, int(request.Acknowledge.Checkpoints), replStream); err != nil {
+				return fmt.Errorf("error handling acknowledgement: %w", err)
+			}
+		default:
+			return fmt.Errorf("unexpected message %#v", request)
+		}
+	}
+}
+
+func (c *Capture) handleAcknowledgement(ctx context.Context, count int, replStream ReplicationStream) error {
+	c.pending.Lock()
+	defer c.pending.Unlock()
+	if count == 0 {
+		return nil // Zero count isn't permitted in the protocol, but has a sensible meaning of 'ignore it' so let's do that
+	}
+	if count > len(c.pending.cursors) {
+		return fmt.Errorf("invalid acknowledgement count %d, only %d pending checkpoints", count, len(c.pending.cursors))
+	}
+
+	var cursor = c.pending.cursors[count-1]
+	c.pending.cursors = c.pending.cursors[count:]
```

Comment:
Does this allow the memory for everything up until `count` to be freed from the backing array? I'm thinking since they are strings and not pointers...maybe? But I've never been 100% sure about. this myself.

Reply:
TL;DR: the old cursor values will eventually get freed

Longer answer with rambling about garbage collection and runtime internals:

Not immediately (because the slice is essentially just a 3-tuple of `(BackingArrayPtr, Length, Capacity)` and `[n:]` means `(BackingArrayPtr+n, Length-n, Capacity-n)` and AFAIK none of the big production-quality garbage collectors do anything but retain a whole allocation when there's an interior pointer). But later on when we `append()` to the resliced array we will eventually run out of capacity and it'll get reallocated and the retained elements copied, and at that point the entire old array is freed.

So this probably does retain a little bit of extra memory, but a replication cursor looks like `"0/E8E2DF0"` so we're talking about kilobytes at most so I'm not inclined to write a more complicated piece of logic here which I'm more likely to get wrong.

Reply:
I see! Thanks for the explanation.